hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightSearchEndpoint.java

```diff
@@ -0,0 +1,457 @@
+import static org.apache.hadoop.ozone.recon.ReconConstants.RECON_OPEN_KEY_INCLUDE_FSO;
+import static org.apache.hadoop.ozone.recon.ReconConstants.RECON_OPEN_KEY_INCLUDE_NON_FSO;
+import static org.apache.hadoop.ozone.recon.api.handlers.BucketHandler.getBucketHandler;
+import static org.apache.hadoop.ozone.recon.api.handlers.EntityHandler.normalizePath;
+import static org.apache.hadoop.ozone.recon.api.handlers.EntityHandler.parseRequestPath;
+
+/**
+ * REST endpoint for search implementation in OM DB Insight.
+ */
+@Path("/insights")
+@Produces(MediaType.APPLICATION_JSON)
+@AdminOnly
+public class OMDBInsightSearchEndpoint {
+
+  private OzoneStorageContainerManager reconSCM;
+  private final ReconOMMetadataManager omMetadataManager;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMDBInsightSearchEndpoint.class);
+  private ReconNamespaceSummaryManagerImpl reconNamespaceSummaryManager;
+
+
+  @Inject
+  public OMDBInsightSearchEndpoint(OzoneStorageContainerManager reconSCM,
+                                   ReconOMMetadataManager omMetadataManager,
+                                   ReconNamespaceSummaryManagerImpl reconNamespaceSummaryManager) {
+    this.reconSCM = reconSCM;
+    this.omMetadataManager = omMetadataManager;
+    this.reconNamespaceSummaryManager = reconNamespaceSummaryManager;
+  }
+
+
+  /**
+   * Performs a search for open keys in the Ozone Manager (OM) database using a specified search prefix.
+   * This endpoint can search across both File System Optimized (FSO) and Object Store (non-FSO) layouts,
+   * compiling a list of keys that match the given prefix along with their data sizes.
+   *
+   * The search prefix may range from the root level ('/') to any specific directory
+   * or key level (e.g., '/volA/' for everything under 'volA'). The search operation matches
+   * the prefix against the start of keys' names within the OM DB.
+   *
+   * Example Usage:
+   *  1. A searchPrefix of "/" will return all keys in the database.
+   *  2. A searchPrefix of "/volA/" retrieves every key under volume 'volA'.
+   *  3. Specifying "/volA/bucketA/dir1" focuses the search within 'dir1' inside 'bucketA' of 'volA'.
+   *
+   * @param searchPrefix The prefix for searching keys, starting from the root ('/') or any specific path.
+   * @param includeFso Indicates whether to include FSO layout keys in the search.
+   * @param includeNonFso Indicates whether to include non-FSO layout keys in the search.
+   * @param limit Limits the number of returned keys.
+   * @return A KeyInsightInfoResponse, containing matching keys and their data sizes.
+   * @throws IOException On failure to access the OM database or process the operation.
+   */
+  @GET
+  @Path("/openKeys/search")
+  public Response searchOpenKeys(
+      @QueryParam("searchPrefix")
+      String searchPrefix,
+      @DefaultValue(DEFAULT_OPEN_KEY_INCLUDE_FSO) @QueryParam(RECON_OPEN_KEY_INCLUDE_FSO)
+      boolean includeFso,
```

Comment:
I included this feature because I believe that if the user provides the prefix `"/"` or `"/volumeName"`, they should have the freedom to choose between `fso` and `non-fso` buckets. If they only wish to view the fso-keys, they can simply set the filter to `true`; otherwise, they can set it to `false`. This approach significantly minimizes resource consumption by avoiding searches for key types not specified by the user.

And the UI for this that already exists shows only one of the keys either `fso` or `non-fso` hence these are the reasons why I strongly believe we should include these filters to the search endpoint.

Reply:
Functionally, it is ok, but we should think of business use case here and why this global backend search is being added as part of this PR. In current UI of open keys, that also provides both (FSO and non-FSO) because on existing current page, user can do UI search of open key and if in that page, key not found, user can move to next page and then again do UI search for that specific open Key, As we can see this is the pain point where user has to keep moving to next pages untill he finds out the specific open key and details of it. Even if we want to include this parameter of FSO or non-FSO in global search, then we should by default set this parameter to "`all`" because this will help user to get the results quickly else he need to do hit and trial again, once with FSO and if not found, then again non-FSO search hit. Rocks DB iterations are quite performant and adding `all` should not have any significant impact on performance.